# <center>Pandas Demo 1 </center>

<img src="img/pandas.jpg" width="250" height="250" />


### Pandas is a fast, powerful, flexible and easy to use open source data analysis and manipulation tool, built on top of the Python programming language.

## Quick Intro



1. Pandas is quite a game changer when it comes to analyzing data with Python and it is one of the most preferred and widely used tools in data munging/wrangling if not THE most used one.


2. Pandas  takes data (like a CSV or TSV file, or a SQL database) and creates a Python object with rows and columns called data frame that looks very similar to table in a statistical software.


3. This is so much easier to work with in comparison to working with lists and/or dictionaries through for loops or list comprehension.

### The pd is a very popular alias given to pandas

In [4]:
import pandas as pd


# Working with Pandas

When you want to use Pandas for data analysis, you’ll usually use it in one of three different ways:

  1. Convert a Python’s list, dictionary or Numpy array to a Pandas data frame
   
   
  2. Open a local file using Pandas, usually a CSV file, but could also be a delimited text file (like TSV), Excel, etc
  
  
 3. Open a remote file or database like a CSV or a JSONon a website through a URL or read from a SQL table/database



#### There are different commands to each of these options, but when you open a file, they would look like this:

##### pd.read_filetype()
for e.g 

 pd.read_csv()




In [ ]:
pd.read_csv('developer_survey_2019/survey_results_public.csv')


# Datastructure in Pandas

There are two types of data structures in pandas: DataFrames and Series.
### DataFrame:
A pandas DataFrame is a two (or more) dimensional data structure – basically a table with rows and columns. The columns have names and the rows have indexes.


In [29]:
df = pd.read_csv('developer_survey_2019/survey_results_public.csv')


In [ ]:
df

### Series
A pandas Series is a one dimensional data structure  that can store values — and for every value it holds a unique index, too.

In [30]:
series = df['Country']


In [ ]:
series

# Viewing And Inspecting Data

Now that you’ve loaded your data, it’s time to take a look. How does the data frame look? Running the name of the data frame would give you the entire table.

You can view the first n rows using df.head(n)

By default, n=5.


In [13]:
df.head()

Respondent                                         MainBranch Hobbyist  \
0           1             I am a student who is learning to code      Yes   
1           2             I am a student who is learning to code       No   
2           3  I am not primarily a developer, but I write co...      Yes   
3           4                     I am a developer by profession       No   
4           5                     I am a developer by profession      Yes   

                  OpenSourcer  \
0                       Never   
1     Less than once per year   
2                       Never   
3                       Never   
4  Once a month or more often   

                                          OpenSource  \
0  The quality of OSS and closed source software ...   
1  The quality of OSS and closed source software ...   
2  The quality of OSS and closed source software ...   
3  The quality of OSS and closed source software ...   
4  OSS is, on average, of HIGHER quality than pro...   

                               Employment                 Country  \
0  Not employed, and not looking for work          United Kingdom   
1      Not employed, but looking for work  Bosnia and Herzegovina   
2                      Employed full-time                Thailand   
3                      Employed full-time           United States   
4                      Employed full-time                 Ukraine   

          Student                                            EdLevel  \
0              No                          Primary/elementary school   
1  Yes, full-time  Secondary school (e.g. American high school, G...   
2              No           Bachelor’s degree (BA, BS, B.Eng., etc.)   
3              No           Bachelor’s degree (BA, BS, B.Eng., etc.)   
4              No           Bachelor’s degree (BA, BS, B.Eng., etc.)   

                                      UndergradMajor  ...  \
0                                                NaN  ...   
1                                                NaN  ...   
2                      Web development or web design  ...   
3  Computer science, computer engineering, or sof...  ...   
4  Computer science, computer engineering, or sof...  ...   

                             WelcomeChange  \
0  Just as welcome now as I felt last year   
1  Just as welcome now as I felt last year   
2  Just as welcome now as I felt last year   
3  Just as welcome now as I felt last year   
4  Just as welcome now as I felt last year   

                                        SONewContent   Age Gender Trans  \
0  Tech articles written by other developers;Indu...  14.0    Man    No   
1  Tech articles written by other developers;Indu...  19.0    Man    No   
2  Tech meetups or events in your area;Courses on...  28.0    Man    No   
3  Tech articles written by other developers;Indu...  22.0    Man    No   
4  Tech meetups or events in your area;Courses on...  30.0    Man    No   

                 Sexuality                                 Ethnicity  \
0  Straight / Heterosexual                                       NaN   
1  Straight / Heterosexual                                       NaN   
2  Straight / Heterosexual                                       NaN   
3  Straight / Heterosexual              White or of European descent   
4  Straight / Heterosexual  White or of European descent;Multiracial   

  Dependents           SurveyLength                  SurveyEase  
0         No  Appropriate in length  Neither easy nor difficult  
1         No  Appropriate in length  Neither easy nor difficult  
2        Yes  Appropriate in length  Neither easy nor difficult  
3         No  Appropriate in length                        Easy  
4         No  Appropriate in length                        Easy  

[5 rows x 85 columns]

In [14]:
df.head(3)

Respondent                                         MainBranch Hobbyist  \
0           1             I am a student who is learning to code      Yes   
1           2             I am a student who is learning to code       No   
2           3  I am not primarily a developer, but I write co...      Yes   

               OpenSourcer                                         OpenSource  \
0                    Never  The quality of OSS and closed source software ...   
1  Less than once per year  The quality of OSS and closed source software ...   
2                    Never  The quality of OSS and closed source software ...   

                               Employment                 Country  \
0  Not employed, and not looking for work          United Kingdom   
1      Not employed, but looking for work  Bosnia and Herzegovina   
2                      Employed full-time                Thailand   

          Student                                            EdLevel  \
0              No                          Primary/elementary school   
1  Yes, full-time  Secondary school (e.g. American high school, G...   
2              No           Bachelor’s degree (BA, BS, B.Eng., etc.)   

                  UndergradMajor  ...  \
0                            NaN  ...   
1                            NaN  ...   
2  Web development or web design  ...   

                             WelcomeChange  \
0  Just as welcome now as I felt last year   
1  Just as welcome now as I felt last year   
2  Just as welcome now as I felt last year   

                                        SONewContent   Age Gender Trans  \
0  Tech articles written by other developers;Indu...  14.0    Man    No   
1  Tech articles written by other developers;Indu...  19.0    Man    No   
2  Tech meetups or events in your area;Courses on...  28.0    Man    No   

                 Sexuality Ethnicity Dependents           SurveyLength  \
0  Straight / Heterosexual       NaN         No  Appropriate in length   
1  Straight / Heterosexual       NaN         No  Appropriate in length   
2  Straight / Heterosexual       NaN        Yes  Appropriate in length   

                   SurveyEase  
0  Neither easy nor difficult  
1  Neither easy nor difficult  
2  Neither easy nor difficult  

[3 rows x 85 columns]

#  .

Similarly, you can view the last n rows using df.tail(n)

Again,by default, n=5.

In [16]:
df.tail()


Respondent MainBranch Hobbyist  \
88878       88377        NaN      Yes   
88879       88601        NaN       No   
88880       88802        NaN       No   
88881       88816        NaN       No   
88882       88863        NaN      Yes   

                                             OpenSourcer  \
88878  Less than once a month but more than once per ...   
88879                                              Never   
88880                                              Never   
88881                                              Never   
88882                            Less than once per year   

                                              OpenSource  \
88878  The quality of OSS and closed source software ...   
88879  The quality of OSS and closed source software ...   
88880                                                NaN   
88881  OSS is, on average, of HIGHER quality than pro...   
88882  OSS is, on average, of HIGHER quality than pro...   

                                              Employment Country  \
88878             Not employed, and not looking for work  Canada   
88879                                                NaN     NaN   
88880                                 Employed full-time     NaN   
88881  Independent contractor, freelancer, or self-em...     NaN   
88882             Not employed, and not looking for work   Spain   

              Student                             EdLevel  \
88878              No           Primary/elementary school   
88879             NaN                                 NaN   
88880             NaN                                 NaN   
88881             NaN                                 NaN   
88882  Yes, full-time  Professional degree (JD, MD, etc.)   

                                          UndergradMajor  ...  \
88878                                                NaN  ...   
88879                                                NaN  ...   
88880                                                NaN  ...   
88881                                                NaN  ...   
88882  Computer science, computer engineering, or sof...  ...   

                                  WelcomeChange  \
88878                                       NaN   
88879                                       NaN   
88880                                       NaN   
88881                                       NaN   
88882  Somewhat less welcome now than last year   

                                            SONewContent   Age Gender Trans  \
88878  Tech articles written by other developers;Tech...   NaN    Man    No   
88879                                                NaN   NaN    NaN   NaN   
88880                                                NaN   NaN    NaN   NaN   
88881                                                NaN   NaN    NaN   NaN   
88882  Tech articles written by other developers;Indu...  18.0    Man    No   

                     Sexuality  \
88878                      NaN   
88879                      NaN   
88880                      NaN   
88881                      NaN   
88882  Straight / Heterosexual   

                                               Ethnicity Dependents  \
88878                                                NaN         No   
88879                                                NaN        NaN   
88880                                                NaN        NaN   
88881                                                NaN        NaN   
88882  Hispanic or Latino/Latina;White or of European...         No   

                SurveyLength SurveyEase  
88878  Appropriate in length       Easy  
88879                    NaN        NaN  
88880                    NaN        NaN  
88881                    NaN        NaN  
88882  Appropriate in length       Easy  

[5 rows x 85 columns]

In [17]:
df.tail(2)

Respondent MainBranch Hobbyist              OpenSourcer  \
88881       88816        NaN       No                    Never   
88882       88863        NaN      Yes  Less than once per year   

                                              OpenSource  \
88881  OSS is, on average, of HIGHER quality than pro...   
88882  OSS is, on average, of HIGHER quality than pro...   

                                              Employment Country  \
88881  Independent contractor, freelancer, or self-em...     NaN   
88882             Not employed, and not looking for work   Spain   

              Student                             EdLevel  \
88881             NaN                                 NaN   
88882  Yes, full-time  Professional degree (JD, MD, etc.)   

                                          UndergradMajor  ...  \
88881                                                NaN  ...   
88882  Computer science, computer engineering, or sof...  ...   

                                  WelcomeChange  \
88881                                       NaN   
88882  Somewhat less welcome now than last year   

                                            SONewContent   Age Gender Trans  \
88881                                                NaN   NaN    NaN   NaN   
88882  Tech articles written by other developers;Indu...  18.0    Man    No   

                     Sexuality  \
88881                      NaN   
88882  Straight / Heterosexual   

                                               Ethnicity Dependents  \
88881                                                NaN        NaN   
88882  Hispanic or Latino/Latina;White or of European...         No   

                SurveyLength SurveyEase  
88881                    NaN        NaN  
88882  Appropriate in length       Easy  

[2 rows x 85 columns]

# .
You can also view n random rows of the data frame using sample method.
##### Note: Here the default value of n is not 5, it is 1.

In [21]:
df.sample()

Respondent                      MainBranch Hobbyist  \
22636       22760  I am a developer by profession      Yes   

                                             OpenSourcer  \
22636  Less than once a month but more than once per ...   

                                              OpenSource          Employment  \
22636  The quality of OSS and closed source software ...  Employed full-time   

              Country Student                                   EdLevel  \
22636  United Kingdom      No  Bachelor’s degree (BA, BS, B.Eng., etc.)   

                                          UndergradMajor  ...  \
22636  A natural science (ex. biology, chemistry, phy...  ...   

                                  WelcomeChange SONewContent   Age Gender  \
22636  Somewhat less welcome now than last year          NaN  28.0    Man   

      Trans                Sexuality                     Ethnicity Dependents  \
22636    No  Straight / Heterosexual  White or of European descent         No   

      SurveyLength SurveyEase  
22636     Too long       Easy  

[1 rows x 85 columns]

In [22]:
df.sample(5)

Respondent                                         MainBranch Hobbyist  \
57143       57506             I am a student who is learning to code      Yes   
44198       44479  I am not primarily a developer, but I write co...      Yes   
88513       31143                                                NaN      Yes   
20386       20500                     I am a developer by profession      Yes   
59861       60245                     I am a developer by profession      Yes   

                                             OpenSourcer  \
57143                                              Never   
44198  Less than once a month but more than once per ...   
88513                         Once a month or more often   
20386                            Less than once per year   
59861  Less than once a month but more than once per ...   

                                              OpenSource          Employment  \
57143  OSS is, on average, of HIGHER quality than pro...  Employed part-time   
44198  OSS is, on average, of HIGHER quality than pro...  Employed full-time   
88513  The quality of OSS and closed source software ...  Employed full-time   
20386  OSS is, on average, of HIGHER quality than pro...  Employed full-time   
59861  The quality of OSS and closed source software ...  Employed full-time   

             Country         Student  \
57143          India  Yes, full-time   
44198  United States              No   
88513  United States  Yes, full-time   
20386         Brazil              No   
59861     Luxembourg             NaN   

                                                 EdLevel  \
57143           Bachelor’s degree (BA, BS, B.Eng., etc.)   
44198  Some college/university study without earning ...   
88513        Master’s degree (MA, MS, M.Eng., MBA, etc.)   
20386           Bachelor’s degree (BA, BS, B.Eng., etc.)   
59861          Other doctoral degree (Ph.D, Ed.D., etc.)   

                                          UndergradMajor  ...  \
57143                      Web development or web design  ...   
44198  Information systems, information technology, o...  ...   
88513  Computer science, computer engineering, or sof...  ...   
20386  Computer science, computer engineering, or sof...  ...   
59861  Computer science, computer engineering, or sof...  ...   

                                 WelcomeChange  \
57143    A lot less welcome now than last year   
44198  Just as welcome now as I felt last year   
88513  Just as welcome now as I felt last year   
20386  Just as welcome now as I felt last year   
59861  Just as welcome now as I felt last year   

                                            SONewContent   Age Gender Trans  \
57143  Tech articles written by other developers;Indu...  21.0    Man    No   
44198  Tech articles written by other developers;Tech...  25.0    Man    No   
88513                                                NaN   NaN    Man    No   
20386  Tech articles written by other developers;Indu...  30.0    Man    No   
59861  Tech articles written by other developers;Indu...  36.0    Man    No   

                     Sexuality  \
57143  Straight / Heterosexual   
44198  Straight / Heterosexual   
88513  Straight / Heterosexual   
20386  Straight / Heterosexual   
59861  Straight / Heterosexual   

                                               Ethnicity Dependents  \
57143                                        South Asian        Yes   
44198  Black or of African descent;White or of Europe...         No   
88513                       White or of European descent         No   
20386                       White or of European descent         No   
59861                       White or of European descent        Yes   

                SurveyLength                  SurveyEase  
57143  Appropriate in length  Neither easy nor difficult  
44198  Appropriate in length  Neither easy nor difficult  
88513  Appropriate in length                        Easy  
20386  Appropriate in length             

### Shape
df.shape() gives you the number of rows and columns of the dataframe respectively.

##### Note: df.shape is attribute not function so there won't be any parenthesis at the end.

In [19]:
df.shape


(88883, 85)

### Info
Info Method  gives you the index, datatype and memory information.
It also tells you about the number of rows and columns.

##### PS: This one is a method not an attribute. 

In [20]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 88883 entries, 0 to 88882
Data columns (total 85 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Respondent              88883 non-null  int64  
 1   MainBranch              88331 non-null  object 
 2   Hobbyist                88883 non-null  object 
 3   OpenSourcer             88883 non-null  object 
 4   OpenSource              86842 non-null  object 
 5   Employment              87181 non-null  object 
 6   Country                 88751 non-null  object 
 7   Student                 87014 non-null  object 
 8   EdLevel                 86390 non-null  object 
 9   UndergradMajor          75614 non-null  object 
 10  EduOther                84260 non-null  object 
 11  OrgSize                 71791 non-null  object 
 12  DevType                 81335 non-null  object 
 13  YearsCode               87938 non-null  object 
 14  Age1stCode              87634 non-null

#### RECALL
When we run the name of the dataframe the whole dataframe prints ...Right.

Let's try run it again


In [28]:
df


Respondent                              MainBranch Hobbyist  \
0               1  I am a student who is learning to code      Yes   
1               2  I am a student who is learning to code       No   
...           ...                                     ...      ...   
88881       88816                                     NaN       No   
88882       88863                                     NaN      Yes   

                   OpenSourcer  \
0                        Never   
1      Less than once per year   
...                        ...   
88881                    Never   
88882  Less than once per year   

                                              OpenSource  \
0      The quality of OSS and closed source software ...   
1      The quality of OSS and closed source software ...   
...                                                  ...   
88881  OSS is, on average, of HIGHER quality than pro...   
88882  OSS is, on average, of HIGHER quality than pro...   

                                              Employment  \
0                 Not employed, and not looking for work   
1                     Not employed, but looking for work   
...                                                  ...   
88881  Independent contractor, freelancer, or self-em...   
88882             Not employed, and not looking for work   

                      Country         Student  \
0              United Kingdom              No   
1      Bosnia and Herzegovina  Yes, full-time   
...                       ...             ...   
88881                     NaN             NaN   
88882                   Spain  Yes, full-time   

                                                 EdLevel  \
0                              Primary/elementary school   
1      Secondary school (e.g. American high school, G...   
...                                                  ...   
88881                                                NaN   
88882                 Professional degree (JD, MD, etc.)   

                                          UndergradMajor  \
0                                                    NaN   
1                                                    NaN   
...                                                  ...   
88881                                                NaN   
88882  Computer science, computer engineering, or sof...   

                                                EduOther OrgSize  \
0      Taught yourself a new language, framework, or ...     NaN   
1      Taken an online course in programming or softw...     NaN   
...                                                  ...     ...   
88881                                                NaN     NaN   
88882  Taken an online course in programming or softw...     NaN   

                                                 DevType YearsCode Age1stCode  \
0                                                    NaN         4         10   
1      Developer, desktop or enterprise applications;...       NaN         17   
...                                                  ...       ...        ...   
88881                                                NaN       NaN        NaN   
88882                                                NaN         8         11   

      YearsCodePro CareerSat JobSat MgrIdiot MgrMoney MgrWant  \
0              NaN       NaN    NaN      NaN      NaN     NaN   
1              NaN       NaN    NaN      NaN      NaN     NaN   
...            ...       ...    ...      ...      ...     ...   
88881          NaN       NaN    NaN      NaN      NaN     NaN   
88882            3       NaN    NaN      NaN      NaN     NaN   

                               JobSeek          LastHireDate LastInt FizzBuzz  \
0                                  NaN                   NaN     NaN      NaN   
1      I am actively looking for a job  I've never had a job     NaN      NaN   
...                                ...                   ...     ...      ...   
88881                          

Wait .. What's that I can't see all 88883 rows on the screen nor can I see all 85 rows on the screen.

# 🤔  🤔  🤔  🤔  🤔

You See, Pandas concatenate the data frame by default.
Now the questions arises ....

How can we view the complete dataset 
# 🤐  🤐  🤐  🤐  🤐
Here is how, 
Pandas have a function set_option
You can explicitly tell the number of rows and columns you want to view...like this


# 🤨  👇  👇  👇 👇  


In [27]:
pd.set_option('max_rows',8,'max_columns',85)

### Now go up and run df again ....there is a surprise waiting

# 🙄  🙄  🙄  🙄  🙄  🙄  🙄


Now play with these function so that you get comfortable with them.
Meet you in the next Video.
Stay Tuned...

<img src="img/stay.jpg" width="250" height="250" />